Introduce React Native OTA updates with zephyr-ota-plugin

HigorAlves opened the pull request:
### What's added in this PR?

This PR introduces the initial implementation of the `zephyr-ota-plugin` library for managing OTA (Over-The-Air) updates in React Native applications. Key features include:

- Core library implementation for seamless OTA updates via Zephyr Cloud.
- API client for fetching remote metadata and updates.
- React hooks and providers for integration into React Native apps.
- Features like periodic update checks, app reload, and version tracking.
- Utilities for dependency management and platform detection.
- Inclusion of unit tests for key components and functionalities.
- Configuration of ESLint, Jest, and TypeScript.
- Comprehensive documentation, example usage code, and default configurations.

#### Screenshots

> _Not applicable._

### What's the issues or discussion related to this PR ?

This library was created to streamline the integration of OTA updates into React Native apps, simplifying version management and minimizing user disruption during updates.

### What are the steps to test this PR?

1. Install the library in a React Native project.
2. Configure the API client with valid credentials.
3. Test periodic update checks, version tracking, and reloading functionality.
4. Run unit tests to ensure utilities and integrations work as expected.

### Documentation update for this PR (if applicable)?

The library's setup guide and usage examples are included in the documentation. Please include this PR into [zephyr-documentation](https://github.com/ZephyrCloudIO/zephyr-documentation).

### (Optional) What's left to be done for this PR?

Fine-tuning and feedback-based iterations may be required after wider usage.

### (Optional) What's the potential risk and how to mitigate it?

Potential configuration issues or platform-specific edge cases. Mitigation includes thorough testing across platforms and providing fallback mechanisms.

### (Required) Pre-PR/Merge checklist

- [ ] I have added/updated/opened a PR to [documentation](https://github.com/ZephyrCloudIO/zephyr-documentation) to cover this new behavior
- [ ] I have added an explanation of my changes
- [ ] I have written new tests (if applicable)
- [ ] I have tested this locally (standing from a first-time user point of view, never touched this app before)
- [ ] I have/will run tests, or ask for help to add test

HigorAlves committed:
feat(zephyr-ota-plugin): introduce new library for React Native OTA updates

- Added core implementation of the `zephyr-ota-plugin` library.
- Integrated API client for resolving remote metadata from Zephyr Cloud.
- Included React hooks and provider for easy integration into React Native applications.
- Supported periodic update checks, app reloads, and version tracking.
- Implemented utilities for dependency parsing and platform detection.
- Added comprehensive unit tests for key utilities and API integration.
- Configured ESLint, Jest, and TypeScript for the library.
- Provided documentation, examples, and default configurations.

HigorAlves committed:
chore: auto detects remotes from the module federation runtime

arthurfiorette commented on libs/zephyr-ota-plugin/src/lib/react/useZephyrOTA.ts:
```diff
@@ -0,0 +1,30 @@
+ * @returns OTA context value with state and actions
+ */
+export function useZephyrOTA(): ZephyrOTAContextValue {
+  return useZephyrOTAContext();
```
whats the point of this fn? why not `useZephyrOTAContext`

HigorAlves committed:
Merge branch 'main' into feat/OTA-updates

# Conflicts:
#	pnpm-lock.yaml

arthurfiorette commented on libs/zephyr-ota-plugin/src/lib/utils/logger.ts:
```diff
@@ -0,0 +1,64 @@
+}
+
+/** Create a scoped logger with a specific prefix */
+export function createScopedLogger(scope: string): {
```
maybe we should use the same logging structure as other plugins? 

`libs/zephyr-agent/src/lib/logging/debug.ts`?

HigorAlves committed:
refactor: separated types from main file and removed useless useMemo

HigorAlves committed:
refactor(zephyr-ota-plugin): simplify imports and reformat comments for clarity

arthurfiorette commented:
Some of them are nitpicks and you can just ignore and some are valid questions.

Nice work btw

arthurfiorette commented on libs/zephyr-ota-plugin/src/lib/core/ZephyrOTAService.ts:
```diff
@@ -0,0 +1,392 @@
+
+    for (const dep of parsedDeps) {
+      logger.debug(`Resolving initial version for ${dep.name}...`);
+      const resolved = await this.apiClient.resolveRemote(dep);
```
Can this work inside a Promise.all?

ryok90 reviewed (commented):
As a ReactProvider this makes sense since it allows the user to input behavior into the OTA responses.
Since this is tightly related to ReactNative and Metro, it would be interesting to see this into metro plugin or repack plugin (if it supports and with a reexport in metro) instead of a separate NPM package.

One thing I noticed on metro plugin is the verbose declaration of options in the withZephyr when we do have all of that in zephyr-engine being derived from users config in MF plugin and other configs such as, target, remotes, name, and all of that. We don't need such parameters since we are already grabbing that info from configs. OTAProvider also won't need those configs if we have access to zephyr-manifest.json.

All of these changes can be done after this PR is merged but for now what is really important is that we reduce user's boilerplate into making OTA work, try and move this Provider into metro plugin, we can document and work on other changes on the next iteration.

HigorAlves committed:
refactor(zephyr-ota-plugin): improve concurrency handling, platform validation, and imports

- Added `mapWithConcurrency` helper to limit concurrent API requests.
- Introduced platform validation to restrict usage to iOS and Android.
- Simplified imports and code structure by removing unused exports and redundant functions.
- Updated `apiBaseUrl` to dynamically resolve from `zephyr-edge-contract`.
- Enhanced package export definitions in `package.json`.

HigorAlves commented:
@claude please review

HigorAlves committed:
Merge branch 'main' into feat/OTA-updates

# Conflicts:
#	pnpm-lock.yaml

HigorAlves committed:
feat(zephyr-ota-plugin): add storage error handling and integrate debug logging

- Introduced `StorageError` and `StorageErrorHandler` types for robust error reporting in storage operations.
- Added `reportError` utility for centralized error handling across storage methods.
- Integrated `debug` package for improved, scoped logging with pre-defined namespaces.
- Updated async storage operations with error reporting callbacks.
- Enhanced `logger` utility to support scoped loggers and dynamic debug enabling/disabling.
- Updated dependencies to include `debug` and `@types/debug`.

HigorAlves committed:
Merge branch 'main' into feat/OTA-updates

HigorAlves committed:
feat(zephyr-ota-plugin): replace concurrency-limited requests with batch resolve API

- Introduced `BatchResolveRequest` and `BatchResolveResponse` types.
- Implemented `resolveRemotesBatch` method in the API client for handling batch resolve requests.
- Replaced `mapWithConcurrency` logic with batch resolve in version tracking and remote resolution.
- Improved error handling with fallback to individual requests when batch resolve fails.

HigorAlves committed:
Merge remote-tracking branch 'origin/feat/OTA-updates' into feat/OTA-updates

HigorAlves committed:
Merge branch 'main' into feat/OTA-updates

HigorAlves committed:
refactor(zephyr-ota-plugin): remove legacy remotes detection and parsing utilities

- Deleted `detect-remotes` and `parse-dependencies` modules along with their tests.
- Removed unused exports and related types from the library.
- Updated `ZephyrOTAProvider` to adopt the new manifest-based update detection flow.
- Streamlined the library API to align with the updated architecture.

HigorAlves committed:
Merge remote-tracking branch 'origin/feat/OTA-updates' into feat/OTA-updates
